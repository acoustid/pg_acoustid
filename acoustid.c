/* acoustid.c */

#include "acoustid.h"


#include "utils/array.h"
#include "catalog/pg_type.h"

PG_MODULE_MAGIC;

/*
PG_FUNCTION_INFO_V1(acoustid_fingerprint_from_int4array);

Datum acoustid_fingerprint_from_int4array(PG_FUNCTION_ARGS) {
    int version = PG_GETARG_INT32(0);
    ArrayType *input = PG_GETARG_ARRAYTYPE_P(1);

    size_t num_terms;
    uint32_t *terms;

    if (ARR_ELEMTYPE(input) != INT4OID) {
        ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                        errmsg("array must contain int4 values")));
    }

    if (ARR_NDIM(input) != 1) {
        ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                        errmsg("array must be one-dimensional")));
    }

    if (ARR_HASNULL(input)) {
        ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                        errmsg("array must not contain nulls")));
    }

    num_terms = ARR_DIMS(input)[0];
    terms = (uint32_t *)ARR_DATA_PTR(input);

    PG_RETURN_FINGERPRINT_P(create_fingerprint(version, num_terms, terms));
}
*/
