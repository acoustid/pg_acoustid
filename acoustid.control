# acoustid extension
comment = 'AcoustID utility functions'
default_version = '2.0'
module_pathname = '$libdir/acoustid'
relocatable = true
