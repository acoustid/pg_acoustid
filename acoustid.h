#ifndef PG_ACOUSTID_H_
#define PG_ACOUSTID_H_

#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <memory.h>

#ifndef NO_PG_EXTENSION
#include "postgres.h"
#include "fmgr.h"
#include "utils/expandeddatum.h"
#include "libpq/pqformat.h"
#else
#include "mock_postgres.h"
#endif

#include "fingerprint.h"

/*
 * Packed fingerprint, implemented as PostgreSQL varlena object.
 *
 * The first 4 bytes contain the total object size in bytes. Be sure to use
 * VARSIZE() and SET_VARSIZE() to access it.
 *
 * The remaining bytes contain the fingerprint data.
 */
typedef struct {
    /* varlena header (do not touch directly!) */
    char vl_len_[4];

    /* fingerprint data */
    uint8_t data[FLEXIBLE_ARRAY_MEMBER];
} FingerprintType;

/*
 * An expanded fingerprint is contained within a private memory context (as
 * all expanded objects must be) and has a control structure as below.
 *
 * The expanded fingerprint might contain a regular packed fingerprint if
 * that was the original input. Otherwise, it contains the fingerprint data
 * as an array of uint32_t values.
 */
#define EP_MAGIC 922322027

typedef struct ExpandedFingerprintHeader {
     /* Standard header for expanded objects */
     ExpandedObjectHeader hdr;

     /*
      * flat_size is the current space requirement for the flat equivalent of
      * the expanded array, if known; otherwise it's 0.  We store this to make
      * consecutive calls of get_flat_size cheap.
      */
     Size flat_size;

     /*
      * fdata points to the flat representation if it is valid, else it is
      * NULL.   
      */
     FingerprintType *fdata;

     /*
      * data is the expanded fingerprint data.
      */
     FingerprintData *data;
} ExpandedFingerprintHeader;

typedef union AnyFingerprintType {
    FingerprintType fp;
    ExpandedFingerprintHeader efp;
} AnyFingerprintType;

#define PG_GETARG_FINGERPRINT_P(x) ((FingerprintType *) PG_DETOAST_DATUM(PG_GETARG_DATUM(x)))
#define PG_RETURN_FINGERPRINT_P(x) PG_RETURN_POINTER(x)

Datum acoustid_fingerprint_type_in(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_type_out(PG_FUNCTION_ARGS);

Datum acoustid_fingerprint_encode_to_text(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_encode_to_bytea(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_decode_from_text(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_decode_from_bytea(PG_FUNCTION_ARGS);

#endif  // PG_ACOUSTID_H_
