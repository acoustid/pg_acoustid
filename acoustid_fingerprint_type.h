#ifndef PG_ACOUSTID_FINGERPRINT_TYPE_H_
#define PG_ACOUSTID_FINGERPRINT_TYPE_H_

#include "postgres.h"
#include "fmgr.h"
#include "lib/stringinfo.h"

typedef struct {
    /* varlena header (do not touch directly!) */
    char vl_len_[4];

    /* fingerprint terms */
    uint32 terms[FLEXIBLE_ARRAY_MEMBER];
} Fingerprint;

/* fmgr interface macros */
#define DatumGetFingerprintP(x) ((Fingerprint *)PG_DETOAST_DATUM(x))
#define PG_GETARG_FINGERPRINT_P(x) DatumGetFingerprintP(PG_GETARG_DATUM(x))
#define PG_RETURN_FINGERPRINT_P(x) PG_RETURN_POINTER(x)

/* fingerprint accessor macros */
#define FINGERPRINT_SIZE(nterms) (VARHDRSZ + sizeof(uint32) * (nterms))
#define FINGERPRINT_NTERMS(fp) ((VARSIZE(fp) - VARHDRSZ) / sizeof(uint32))
#define FINGERPRINT_TERM(fp, i) ((fp)->terms[i])

Fingerprint *create_empty_fingerprint(int nterms);

Fingerprint *create_fingerprint_from_str(char *str);
void fingerprint_to_str(Fingerprint *fp, StringInfo buf);

Fingerprint *create_fingerprint_from_bytes(StringInfo buf);
void fingerprint_to_bytes(Fingerprint *fp, StringInfo buf);

Datum acoustid_fingerprint_in(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_out(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_recv(PG_FUNCTION_ARGS);
Datum acoustid_fingerprint_send(PG_FUNCTION_ARGS);

#endif // PG_ACOUSTID_FINGERPRINT_TYPE_H_
