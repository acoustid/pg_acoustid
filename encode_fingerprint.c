#include "encode_fingerprint.h"

#include "pack_int3_array.h"
#include "pack_int5_array.h"
#include "unpack_int3_array.h"
#include "unpack_int5_array.h"
#include "uint8_vector.h"

static const int MAX_NORMAL_BIT_VALUE = (1 << 3) - 1;

void encode_fingerprint(FingerprintData *fp, uint8_t **output, size_t *output_len, size_t extra_header) {
    uint8_t *ptr;
    uint32_t last_term;
    int i, num_terms;

    UInt8Vector normal_bits;
    UInt8Vector exceptional_bits;

    num_terms = fp->size;

    uint8_vector_alloc(&normal_bits, num_terms);
    uint8_vector_alloc(&exceptional_bits, num_terms);

    last_term = 0;
    for (i = 0; i < num_terms; i++) {
        int bit = 1, last_bit = 0;
        uint32_t term = fp->terms[i];
        uint32_t term_bits = term ^ last_term;
        last_term = term;
        while (term_bits) {
            if ((term_bits & 1) != 0) {
                int value = bit - last_bit;
                if (value >= MAX_NORMAL_BIT_VALUE) {
                    uint8_vector_push_back(&normal_bits, MAX_NORMAL_BIT_VALUE);
                    uint8_vector_push_back(&exceptional_bits, value - MAX_NORMAL_BIT_VALUE);
                } else {
                    uint8_vector_push_back(&normal_bits, value);
                }
                last_bit = bit;
            }
            term_bits >>= 1;
            bit++;
        }
        uint8_vector_push_back(&normal_bits, 0);
    }

    *output_len = 4 + GetPackedInt3ArraySize(normal_bits.size) + GetPackedInt5ArraySize(exceptional_bits.size);
    *output = palloc0(*output_len + extra_header);

    ptr = *output + extra_header;

    ptr[0] = fp->version;
    ptr[1] = num_terms >> 16;
    ptr[2] = num_terms >> 8;
    ptr[3] = num_terms;
    ptr += 4;

    ptr = PackInt3Array(normal_bits.data, normal_bits.data + normal_bits.size, ptr);
    ptr = PackInt5Array(exceptional_bits.data, exceptional_bits.data + exceptional_bits.size, ptr);

    uint8_vector_free(&normal_bits);
    uint8_vector_free(&exceptional_bits);

    *output_len = ptr - (*output + extra_header);
}

FingerprintData *decode_fingerprint(const uint8_t *input, size_t input_len) {
    FingerprintData *fp;
    int i, j, version, num_terms, num_bits, num_exceptional_bits, found_terms, bit, last_bit, cursor = 0;
    uint32_t term, last_term;
    const uint8_t *header;
    uint8_t *bits, *exceptional_bits;

    if (cursor + 4 > input_len) {
        elog(ERROR, "Invalid fingerprint (shorter than 4 bytes)");
        return NULL;
    }

    // Read the header
    header = input + cursor;
    version = header[0];
    num_terms = ((uint32_t)header[1] << 16) | ((uint32_t)header[2] << 8) | header[3];
    cursor += 4;

    // Estimate the number of bits stored in the packed fingerprint
    num_bits = GetUnpackedInt3ArraySize(input_len - cursor);
    bits = palloc(num_bits * sizeof(uint8_t));

    // Unpack the normal bits
    UnpackInt3Array(input + cursor, input + input_len, bits);

    // Find the actual number of normal bits, count the number of exceptional
    // bits
    found_terms = 0;
    num_exceptional_bits = 0;
    for (i = 0; i < num_bits; i++) {
        if (bits[i] == 0) {
            found_terms++;
            if (found_terms == num_terms) {
                num_bits = i + 1;
                break;
            }
        } else if (bits[i] == MAX_NORMAL_BIT_VALUE) {
            num_exceptional_bits++;
        }
    }

    // Advance the cursor to the end of the normal bits
    cursor += GetPackedInt3ArraySize(num_bits);

    if (found_terms != num_terms) {
        pfree(bits);
        elog(ERROR, "Invalid fingerprint (too short, not enough input for "
                    "normal bits)");
        return NULL;
    }

    if (cursor + GetPackedInt5ArraySize(num_exceptional_bits) > input_len) {
        pfree(bits);
        elog(ERROR, "Invalid fingerprint (too short, not enough input for "
                    "exceptional bits)");
        return NULL;
    }

    if (num_exceptional_bits > 0) {
        // Estimate the number of exceptional bits stored in the packed
        // fingerprint
        exceptional_bits =
            palloc(GetUnpackedInt5ArraySize(GetPackedInt5ArraySize(num_exceptional_bits)) * sizeof(uint8_t));
        // Unpack the exceptional bits
        UnpackInt5Array(input + cursor, input + input_len, exceptional_bits);
        // Add the exceptional bits to the normal bits
        for (i = 0, j = 0; i < num_bits && j < num_exceptional_bits; i++) {
            if (bits[i] == MAX_NORMAL_BIT_VALUE) {
                bits[i] += exceptional_bits[j++];
            }
        }
        pfree(exceptional_bits);
    }

    fp = fingerprint_data_alloc(num_terms);
    fp->version = version;
    fp->size = num_terms;

    // Unpack the bits into fingerprint terms
    last_term = 0;
    last_bit = 0;
    term = 0;
    for (i = 0, j = 0; i < num_bits && j < num_terms; i++) {
        bit = bits[i];
        if (bit == 0) {
            last_term ^= term;
            fp->terms[j] = last_term;
            last_bit = 0;
            term = 0;
            j++;
        } else {
            bit += last_bit;
            last_bit = bit;
            term |= 1 << (bit - 1);
        }
    }

    pfree(bits);
    return fp;
}
