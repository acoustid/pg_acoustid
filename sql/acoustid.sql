SET client_min_messages = warning;
\set ECHO none
\i acoustid.sql
\set ECHO all
RESET client_min_messages;

SELECT acoustid_compare('{}'::int4[], '{}'::int4[]);
SELECT acoustid_compare('{1}'::int4[], '{}'::int4[]);
SELECT acoustid_compare('{1}'::int4[], '{1}'::int4[]);
SELECT acoustid_compare('{1}'::int4[], '{2}'::int4[]);
SELECT acoustid_compare('{1}'::int4[], '{6}'::int4[]);
SELECT acoustid_compare('{1}'::int4[], '{1,6}'::int4[]);
SELECT acoustid_compare('{1,56}'::int4[], '{1,6}'::int4[]);
SELECT acoustid_compare('{448,56}'::int4[], '{1,6}'::int4[]);

SELECT acoustid_compare('{0,7,56,3584}'::int4[], '{0,7,56,448}'::int4[]);
SELECT acoustid_compare('{7,0,7,56,3584}'::int4[], '{0,7,56,448}'::int4[]);
SELECT acoustid_compare('{0,7,56,3584}'::int4[], '{7,0,7,56,448}'::int4[]);

SELECT acoustid_compare('{65535,65535,1,6,65535,65535}'::int4[], '{1,6}'::int4[]);
SELECT acoustid_compare('{1,6}'::int4[], '{65535,65535,1,6,65535,65535}'::int4[]);

SELECT acoustid_compare('{1,1,1,1,1}'::int4[], '{1,1,1,1,1}'::int4[]);

