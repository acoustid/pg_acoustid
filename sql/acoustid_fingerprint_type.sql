CREATE EXTENSION acoustid;

SELECT '{198178303,198047167,164426175}'::acoustid_fingerprint;
SELECT '{4294967295}'::acoustid_fingerprint;
SELECT '{4294967295}v1'::acoustid_fingerprint;
SELECT '{}'::acoustid_fingerprint;
SELECT '1,2,3'::acoustid_fingerprint;
SELECT '{1,2,3'::acoustid_fingerprint;
SELECT '{a}'::acoustid_fingerprint;
SELECT '1,2,3}'::acoustid_fingerprint;
SELECT '{1, 2, 3}'::acoustid_fingerprint;

CREATE TABLE test_table (
    id SERIAL PRIMARY KEY,
    fingerprint acoustid_fingerprint NOT NULL
);

INSERT INTO test_table (fingerprint) VALUES
    ('{198178303,198047167,164426175}'),
    ('{4294967295}'),
    ('{4294967295}v1'),
    ('{}');

SELECT * FROM test_table;
