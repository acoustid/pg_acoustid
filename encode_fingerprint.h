#ifndef PG_ACOUSTID_ENCODE_FINGERPRINT_H
#define PG_ACOUSTID_ENCODE_FINGERPRINT_H

#include "acoustid.h"
#include "fingerprint.h"

extern void encode_fingerprint(FingerprintData *fp, uint8_t **output, size_t *output_len, size_t extra_header);
extern FingerprintData *decode_fingerprint(const uint8_t *input, size_t input_len);

#endif
