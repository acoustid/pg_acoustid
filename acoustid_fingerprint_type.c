/* acoustid_fingerprint_type.c */

#include "acoustid.h"

#include "encode_fingerprint.h"

PG_FUNCTION_INFO_V1(acoustid_fingerprint_type_in);
PG_FUNCTION_INFO_V1(acoustid_fingerprint_type_out);

Datum acoustid_fingerprint_type_in(PG_FUNCTION_ARGS) {
    char *str = PG_GETARG_CSTRING(0);
    FingerprintData *fp;
    FingerprintType *result;
    uint8_t *encoded_data;
    size_t encoded_size;

    fp = fingerprint_data_from_string(str);
    encode_fingerprint(fp, &encoded_data, &encoded_size, VARHDRSZ);

    result = (FingerprintType *) encoded_data;
    SET_VARSIZE(result, encoded_size + VARHDRSZ);

    PG_RETURN_FINGERPRINT_P(result);
}

Datum acoustid_fingerprint_type_out(PG_FUNCTION_ARGS) {
    FingerprintType *input = PG_GETARG_FINGERPRINT_P(0);
    FingerprintData *fp;
    char *result;
    uint8_t *encoded_data;
    size_t encoded_size;

    encoded_data = (uint8_t *) VARDATA(input);
    encoded_size = VARSIZE(input) - VARHDRSZ;

    fp = decode_fingerprint(encoded_data, encoded_size);
    result = fingerprint_data_to_string(fp);

    PG_RETURN_CSTRING(result);
}
