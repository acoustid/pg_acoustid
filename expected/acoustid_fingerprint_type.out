CREATE EXTENSION acoustid;
SELECT '{198178303,198047167,164426175}'::acoustid_fingerprint;
      acoustid_fingerprint       
---------------------------------
 {198178303,198047167,164426175}
(1 row)

SELECT '{4294967295}'::acoustid_fingerprint;
 acoustid_fingerprint 
----------------------
 {4294967295}
(1 row)

SELECT '{4294967295}v1'::acoustid_fingerprint;
 acoustid_fingerprint 
----------------------
 {4294967295}v1
(1 row)

SELECT '{}'::acoustid_fingerprint;
 acoustid_fingerprint 
----------------------
 {}
(1 row)

SELECT '1,2,3'::acoustid_fingerprint;
ERROR:  Fingerprint string must start with '{'
LINE 1: SELECT '1,2,3'::acoustid_fingerprint;
               ^
SELECT '{1,2,3'::acoustid_fingerprint;
ERROR:  Fingerprint string must end with '}'
LINE 1: SELECT '{1,2,3'::acoustid_fingerprint;
               ^
SELECT '{a}'::acoustid_fingerprint;
ERROR:  Fingerprint string must contain only digits
LINE 1: SELECT '{a}'::acoustid_fingerprint;
               ^
SELECT '1,2,3}'::acoustid_fingerprint;
ERROR:  Fingerprint string must start with '{'
LINE 1: SELECT '1,2,3}'::acoustid_fingerprint;
               ^
SELECT '{1, 2, 3}'::acoustid_fingerprint;
ERROR:  Fingerprint string must not contain spaces
LINE 1: SELECT '{1, 2, 3}'::acoustid_fingerprint;
               ^
CREATE TABLE test_table (
    id SERIAL PRIMARY KEY,
    fingerprint acoustid_fingerprint NOT NULL
);
INSERT INTO test_table (fingerprint) VALUES
    ('{198178303,198047167,164426175}'),
    ('{4294967295}'),
    ('{4294967295}v1'),
    ('{}');
SELECT * FROM test_table;
 id |           fingerprint           
----+---------------------------------
  1 | {198178303,198047167,164426175}
  2 | {4294967295}
  3 | {4294967295}v1
  4 | {}
(4 rows)

